<launch>
  <!-- Mainly copied from -->
  <!-- https://github.com/jsk-ros-pkg/jsk_recognition/blob/master/jsk_pcl_ros/sample/sample_icp_registration.launch -->

  <arg name="gui" default="true"/>
  <arg name="manager" default="tilt_laser_manager"/>

  <!-- input cloud -->
  <include file="$(find kitchen_registration)/launch/tilt_laser_listener.launch">
    <arg name="manager" value="$(arg manager)" />
  </include>

  <node name ="subsample_input" pkg="nodelet" type="nodelet"
      args="load pcl/VoxelGrid $(arg manager)" output="screen">
      <remap from="~input" to="/tilt_laser_listener/output_cloud" />
    <rosparam>
      leaf_size: 0.008
    </rosparam>
  </node>

  <!-- reference cloud -->
  <node name="pointcloud_database_server" pkg="nodelet" type="nodelet"
    args="load jsk_pcl/PointcloudDatabaseServer $(arg manager)" output="screen">
    <rosparam subst_value="true">
      models: ["$(find kitchen_registration)/data/base_link.pcd",
      ]
      duration: 0.05
      use_array: false
    </rosparam>
  </node>

  <node name ="subsample_reference" pkg="nodelet" type="nodelet"
      args="load pcl/VoxelGrid $(arg manager)" output="screen">
    <remap from="~input" to="pointcloud_database_server/cloud" />
    <rosparam>
      leaf_size: 0.005
    </rosparam>
  </node>

  <!-- icp registration -->
  <node name="icp_registration"
        pkg="nodelet" type="nodelet"
        args="load jsk_pcl/ICPRegistration /$(arg manager)"
        output="screen">
    <remap from="~input" to="subsample_input/output" />
    <remap from="~input_reference" to="subsample_reference/output" />
  </node>

  <!-- visualize -->
  <group if="$(arg gui)">
    <node name="rviz"
          pkg="rviz" type="rviz"
          args="-d $(find kitchen_registration)/config/icp_registration.rviz"/>
  </group>
</launch>
